### PR TITLE
Pre-commit hooks for notebook formatting & reproducibility

### PR DESCRIPTION
**Pull request by bdice**
## Description

This PR adds pre-commit checks that should make notebooks a bit more reproducible and improves & automates notebook cell formatting. Metadata is no longer committed, which is a good thing. I verified that the tags discussed in #21 related to nbsphinx hiding certain cells and continuing execution after a cell failure are preserved.

```yml
resources:
  repositories:
    - repository: asottile
      type: github
      endpoint: github
      name: asottile/azure-pipeline-templates
      ref: refs/tags/v2.1.0

jobs:
- template: job--pre-commit.yml@asottile
  parameters:
  - python: "3.8"
```

## Checklist:
- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-examples/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-examples/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of authors](https://github.com/glotzerlab/hoomd-examples/blob/master/AUTHORS.md).

### CODE REVIEW

**Comment by joaander** [comment]
Yes, I've configured an endpoint like that for the hoomd azp config that posts releases. You may be able to reuse the existing endpoint.

**Comment by joaander** [comment]
> Yes, I've configured an endpoint like that for the hoomd azp config that posts releases. You may be able to reuse the existing endpoint.

Or maybe not, they appear to be per project. The ones I've configured are "release" in gsd, hoomd, and fresnel.

**Comment by bdice** [comment]
@joaander The CI check seems to work! It's fine with me if you want to merge this without the "execute notebooks" step, or combine this into #21.

File: .azp/test.yml
Comment (joaander) [inline .azp/test.yml:26]: So we are running a job that is define by a .yml file in asottile/azure-pipeline-templates. What if this repo changes the file, or goes stale? We just copy the file in to our templates directory and remove the resources block to keep control.

File: 00-Introducing-HOOMD-blue/03-Initializing-the-System-State.ipynb
Comment (joaander) [inline 00-Introducing-HOOMD-blue/03-Initializing-the-System-State.ipynb:110]: In my opinion, black makes code blocks like this harder to read.

File: 00-Introducing-HOOMD-blue/03-Initializing-the-System-State.ipynb
Comment (joaander) [inline 00-Introducing-HOOMD-blue/03-Initializing-the-System-State.ipynb:137]: Spaces around the exponent operator make it hard to read.

**Review by joaander — COMMENTED** [review]
Thanks!

Does nbqa work with any formatters other than black?

File: .azp/test.yml
Comment (bdice) [reply .azp/test.yml:26]: It's tagged to a release so this file should not be expected to change. I would prefer to let it use the external template for simplicity. `ref: refs/tags/v2.1.0`

**Comment by bdice** [comment]
> Does nbqa work with any formatters other than black?

Yes, nbqa is flexible. I suspect yapf will work if that's your preference. If you have a yapf settings file or other configuration you want to use, let me know.

**Comment by joaander** [comment]
> > Does nbqa work with any formatters other than black?
> 
> Yes, nbqa is flexible. I suspect yapf will work if that's your preference. If you have a yapf settings file or other configuration you want to use, let me know.

This is the yapf config I use with HOOMD:
```
[yapf]
based_on_style = google
align_closing_bracket_with_visual_indent = True
split_before_arithmetic_operator = True
split_before_bitwise_operator = True
split_before_logical_operator = True
blank_line_before_module_docstring = True
split_before_dot = True
```

You may want to wait until we merge #21 which rewrites many of the notebook files and accept merge those incoming changes before rerunning the formatting.

**Comment by bdice** [comment]
@joaander I updated this PR and I think it's using yapf correctly. I copied the `setup.cfg` directly from HOOMD `master` and created a PR adding yapf to nbQA: https://github.com/nbQA-dev/nbQA/pull/547

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**
